File: modules/xml/app/controllers/xml/XmlController.java
package controllers.xml;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import controllers.core.ExerciseController;
import model.CommonUtils$;
import model.CorrectionException;
import model.StringConsts;
import model.XmlCorrector;
import model.XmlError;
import model.XmlExType;
import model.XmlExercise;
import model.exercise.ExerciseOptions;
import model.result.CompleteResult;
import model.user.User;
import play.Logger;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.mvc.Result;
import play.twirl.api.Html;

public class XmlController extends ExerciseController<XmlExercise, XmlError> {

  private static final ExerciseOptions EX_OPTIONS = new ExerciseOptions("Xml", "xml", 10, 20, false);

  public static final String STANDARD_XML_PLAYGROUND = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
      + "\n<!DOCTYPE root [\n\n]>\n";

  private static final String SAVE_ERROR_MSG = "An error has occured while saving an xml file to ";

  @Inject
  public XmlController(FormFactory theFactory) {
    super(theFactory, "xml", XmlExercise.finder, XmlRoutesObject$.MODULE$);
  }

  public static String getReferenceCode(XmlExercise exercise) {
    final Path referenceFilePath = Paths.get(getSampleDir("xml").toString(),
        exercise.rootNode + "." + exercise.getReferenceFileEnding());
    return referenceFilePath.toFile().exists() ? CommonUtils$.MODULE$.readFile(referenceFilePath) : "FEHLER!";
  }

  public Result playground() {
    return ok(views.html.xmlPlayground.render(getUser()));
  }

  public Result playgroundCorrection() {
    final List<XmlError> result = XmlCorrector.correct(factory.form().bindFromRequest().get(StringConsts.FORM_VALUE),
        "", XmlExType.XML_DTD);
    return ok(views.html.xmlResult.render(result));
  }

  private String readDefOrOldSolution(String username, XmlExercise exercise) {
    final Path oldSolutionPath = getSolFileForExercise(username, exerciseType, exercise, exercise.rootNode,
        exercise.getStudentFileEnding());
    return oldSolutionPath.toFile().exists() ? CommonUtils$.MODULE$.readFile(oldSolutionPath)
        : exercise.getFixedStart();
  }

  private Path saveGrammar(Path dir, String learnerSolution, XmlExercise exercise) {
    final Path grammar = Paths.get(dir.toString(), exercise.rootNode + exercise.getGrammarFileEnding());

    try {
      return Files.write(grammar, Arrays.asList(learnerSolution.split(StringConsts.NEWLINE)), StandardOpenOption.CREATE,
          StandardOpenOption.TRUNCATE_EXISTING);
    } catch (final IOException error) {
      Logger.error("Fehler beim Speichern einer Xml-Loesungsdatei!", error);
      return null;
    }
  }

  private Path saveGrammar(Path dir, XmlExercise exercise) {
    final String filename = exercise.rootNode + "." + exercise.getGrammarFileEnding();

    final Path target = Paths.get(dir.toString(), filename);
    final Path source = Paths.get(getSampleDir().toString(), filename);

    try {
      return Files.copy(source, target, StandardCopyOption.REPLACE_EXISTING);
    } catch (final IOException error) {
      Logger.error(SAVE_ERROR_MSG + target.toString(), error);
      return null;
    }
  }

  private Path saveXML(Path dir, String learnerSolution, XmlExercise exercise) {
    final Path targetFile = Paths.get(dir.toString(), exercise.rootNode + ".xml");

    try {
      return Files.write(targetFile, Arrays.asList(learnerSolution.split(StringConsts.NEWLINE)),
          StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
    } catch (final IOException error) {
      Logger.error(SAVE_ERROR_MSG + targetFile.toString(), error);
      return null;
    }
  }

  private Path saveXML(Path dir, XmlExercise exercise) {
    final String filename = exercise.rootNode + ".xml";

    final Path target = Paths.get(dir.toString(), filename);
    final Path source = Paths.get(getSampleDir().toString(), filename);

    try {
      return Files.copy(source, target, StandardCopyOption.REPLACE_EXISTING);
    } catch (final IOException error) {
      Logger.error(SAVE_ERROR_MSG + target.toString(), error);
      return null;
    }
  }

  @Override
  protected CompleteResult<XmlError> correct(DynamicForm form, XmlExercise exercise, User user) {
    final String learnerSolution = form.get(StringConsts.FORM_VALUE);
    final Path dir = checkAndCreateSolDir(user.name, exercise);

    Path grammar;
    Path xml;
    if(exercise.exerciseType == XmlExType.DTD_XML || exercise.exerciseType == XmlExType.XSD_XML) {
      grammar = saveGrammar(dir, learnerSolution, exercise);
      xml = saveXML(dir, exercise);
    } else {
      grammar = saveGrammar(dir, exercise);
      xml = saveXML(dir, learnerSolution, exercise);
    }

    List<XmlError> res;
    try {
      res = XmlCorrector.correct(xml, grammar, exercise);
    } catch (final CorrectionException e) {
      res = Arrays.asList(new XmlError(null, null));
    }
    return new CompleteResult<>(learnerSolution, res);
  }

  @Override
  protected Html renderExercise(User user, XmlExercise exercise) {
    final String defOrOldSolution = readDefOrOldSolution(user.name, exercise);
    return views.html.exercise2Rows.render(user, XmlRoutesObject$.MODULE$, EX_OPTIONS, exercise,
        views.html.xmlExRest.render(exercise), defOrOldSolution);
  }

  @Override
  protected Html renderExesListRest() {
    return views.html.xmlExesListRest.render();
  }

  @Override
  protected Html renderResult(CompleteResult<XmlError> correctionResult) {
    return views.html.xmlResult.render(correctionResult.getResults());
  }

}

File: modules/xml/app/controllers/xml/XmlController.scala
package controllers.xml

import java.nio.file._
import javax.inject.Inject

import controllers.excontrollers.{AExerciseAdminController, IdExController}
import model.CommonUtils.RicherTry
import model.StringConsts._
import model._
import model.exercise.ExerciseOptions
import model.result.CompleteResult
import model.user.User
import play.api.data.Form
import play.api.mvc.ControllerComponents
import play.twirl.api.{Html, HtmlFormat}

import scala.collection.JavaConverters.{asScalaBufferConverter, seqAsJavaListConverter}
import scala.util.Try

class XmlAdmin @javax.inject.Inject()(cc: ControllerComponents)
  extends AExerciseAdminController[XmlExercise](cc, XmlToolObject, XmlExercise.finder, model.XmlExerciseReader) {

  override def statistics = new Html(
    s"""<li>Es existieren insgesamt ${XmlExercise.finder.all.size} <a href="${controllers.xml.routes.XmlAdmin.exercises()}">Aufgaben</a>, davon
       |  <ul>
       |    ${XmlExercise.finder.all.asScala.groupBy(_.exerciseType).map({ case (exType, exes) => s"""<li>${exes.size} Aufgaben von Typ $exType""" }).mkString("\n")}
       |  </ul>
       |</li>""".stripMargin)
}

class XmlController @Inject()(cc: ControllerComponents)
  extends IdExController[XmlExercise, XmlError](cc, XmlExercise.finder, XmlToolObject) {

  override type SolType = StringSolution

  override def solForm: Form[StringSolution] = ???

  val EX_OPTIONS = ExerciseOptions("Xml", "xml", 10, 20, updatePrev = false)

  val SAVE_ERROR_MSG = "An error has occured while saving an xml file to "

  override protected def correctEx(sol: StringSolution, exercise: XmlExercise, user: User): Try[CompleteResult[XmlError]] =
    checkAndCreateSolDir(user.name, exercise).flatMap(dir => {
      val learnerSolution = sol.learnerSolution

      val (grammarTry, xmlTry) = exercise.exerciseType match {
        case (DTD_XML | XSD_XML) => (
          save(dir, exercise.rootNode + "." + exercise.exerciseType.gramFileEnding, learnerSolution),
          copy(dir, exercise.rootNode + "." + "xml")
        )
        case _ => (
          copy(dir, exercise.rootNode + "." + exercise.exerciseType.gramFileEnding),
          save(dir, exercise.rootNode + "." + "xml", learnerSolution)
        )
      }

      grammarTry.zip(xmlTry).map({ case (grammar, xml) =>
        new CompleteResult(learnerSolution, XmlCorrector.correct(xml, grammar, exercise))
      })
    })


  def playground = Action { implicit request =>
    Ok(views.html.xmlPlayground.render(getUser))
  }

  def playgroundCorrection = Action { implicit request =>
    val sol = request.body.asFormUrlEncoded.get(FORM_VALUE).mkString("\n")
    Ok(renderResult(new CompleteResult("", XmlCorrector.correct(sol, "", XML_DTD))))
  }

  override def renderExercise(user: User, exercise: XmlExercise): Html = views.html.exercise2Rows.render(
    user, toolObject, EX_OPTIONS, exercise, renderExRest(exercise), readDefOrOldSolution(user.name, exercise)
  )

  def renderExRest(exercise: XmlExercise) = new Html(
    s"""<section id="refFileSection">
       |  <pre>${HtmlFormat.escape(getReferenceCode(exercise))}</pre>
       |</section>
""".stripMargin)

  override def renderExesListRest = new Html(
    s"""<div class="panel panel-default">
       |  <a class="btn btn-primary btn-block" href="${controllers.xml.routes.XmlController.playground()}">Xml-Playground</a>
       |</div>

       |<hr>""".stripMargin)

  override def renderResult(completeResult: CompleteResult[XmlError]): Html = completeResult.results match {
    case Nil => new Html("""<div class="alert alert-success">Es wurden keine Fehler gefunden.</div>""")
    case results => new Html(results.map(res =>
      s"""<div class="panel panel-${res.getBSClass}">
         |  <div class="panel-heading">${res.title} ${res.lineStr}</div>
         |  <div class="panel-body">${res.errorMessage}</div>
         |</div>""".stripMargin).mkString("\n"))
  }

  def readDefOrOldSolution(username: String, exercise: XmlExercise): String = Try(
    Files.readAllLines(toolObject.getSolFileForExercise(username, exercise, exercise.rootNode, exercise.exerciseType.studFileEnding)
    ).asScala.mkString("\n")
  ).getOrElse(exercise.fixedStart)

  def copy(dir: Path, filename: String) = Try(Files.copy(Paths.get(toolObject.sampleDir.toString, filename),
    Paths.get(dir.toString, filename), StandardCopyOption.REPLACE_EXISTING))

  def save(dir: Path, filename: String, learnerSolution: String) = Try(
    Files.write(Paths.get(dir.toString, filename), learnerSolution.split(StringConsts.NEWLINE).toList.asJava,
      StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)
  )

  def getReferenceCode(exercise: XmlExercise): String = {
    val referenceFilePath = Paths.get(toolObject.sampleDir.toString, exercise.rootNode + "." + exercise.exerciseType.refFileEnding)
    Try(Files.readAllLines(referenceFilePath).asScala.mkString("\n")).getOrElse("FEHLER!")
  }

}

object XmlController {

  val STANDARD_XML_PLAYGROUND: String =
    """<?xml version="1.0" encoding="utf-8"?>
      |<!DOCTYPE root [
      |
      |]>""".stripMargin

}
